var/spack/repos/builtin/packages/node-js/package.py

```diff
@@ -66,7 +66,10 @@ class NodeJs(Package):
 
     # https://github.com/nodejs/node/blob/master/BUILDING.md#unix-and-macos
     depends_on("gmake@3.81:", type="build")
-    depends_on("python@3.6:3.11", when="@19.1:", type="build")
+    # python requirement is "a supported python version".
+    depends_on("python@3.9:", when="@21:", type="build")
+    # ... even if in the past it was a bit more specific
+    depends_on("python@3.6:3.11", when="@19.1:20", type="build")
     depends_on("python@3.6:3.10", when="@16.11:19.0", type="build")
     depends_on("python@3.6:3.9", when="@16.0:16.10", type="build")
     depends_on("python@2.7,3.5:3.8", when="@15", type="build")
```

Comment:
I wonder if it does not make more sense to split this into separate lower limits (change of minimum required python version as older versions reach EOL and are not supported anymore) and upper limits (new version support introduced). Having this as ranges is leading to repetition.

Reply:
I'm guessing Node.js does that thing where even numbers are LTS releases and odd numbers are development releases? This does indeed make this tricky, although I think it's worthwhile to include detailed upper/lower bounds, even if that means a bit more duplication.

Reply:
@adamjstewart would you go for something like this?

```python
depends_on("python@2.7,3.5:3.7", when="@v13.0.0:v13.0.1", type="build")
depends_on("python@2.7,3.5:3.8", when="@v13.1.0:v14.13.1", type="build")
depends_on("python@2.7,3.5:3.9", when="@v14.14.0:v14.18.1", type="build")
depends_on("python@2.7,3.5:3.10", when="@v14.18.2:v14.21.3", type="build")
depends_on("python@2.7,3.5:3.9", when="@v15.0.0:v15.14.0", type="build")
depends_on("python@3.6:3.9", when="@v16.0.0:v16.10.0", type="build")
depends_on("python@3.6:3.10", when="@v16.11.0:v16.18.1", type="build")
depends_on("python@3.6:3.11", when="@v16.19.0:v16.20.2", type="build")
depends_on("python@3.6:3.10", when="@v17.0.0:v18.12.1", type="build")
depends_on("python@3.6:3.11", when="@v18.13.0:v18.19.1", type="build")
depends_on("python@3.6:3.12", when="@v18.20.0:v18.20.5", type="build")
depends_on("python@3.6:3.10", when="@v19.0.0:v19.0.1", type="build")
depends_on("python@3.6:3.11", when="@v19.1.0:v20.10.0", type="build")
depends_on("python@3.6:3.12", when="@v20.11.0:v20.15.1", type="build")
depends_on("python@3.6:3.13", when="@v20.16.0:v20.18.1", type="build")
depends_on("python@3.6:3.11", when="@v21.0.0:v21.1.0", type="build")
depends_on("python@3.6:3.12", when="@v21.2.0:v22.2.0", type="build")
depends_on("python@3.6:3.13", when="@v22.3.0:v22.7.0", type="build")
```

I created this small script that makes use of the wonderful `spack.version` helpers and provides the output above

```python
#!/usr/bin/env python3

from sys import stdin

if stdin.isatty():
    print("usage: from the root of the repository https://github.com/nodejs/node")
    print(f"git --no-pager grep 'acceptable_pythons =' $(git tag --list | sort -V) -- configure | cut -d: -f1,3 | spack python {__file__}")
    exit(1)

import re
from spack.version import VersionList, VersionRange, ver

data = []
for rawline in stdin.readlines():
    m = re.match(r'(.*):acceptable_pythons = (.*)', rawline)
    if m is None:
        print("No match")

    nodejs_version = ver('=' + m.group(1))

    py_reqs = sorted(eval(m.group(2)))
    py_reqs = [ver('.'.join(map(str, v))) for v in py_reqs]
    py_reqs = VersionList(py_reqs)
    
    data.append((nodejs_version, py_reqs))


last_py_reqs = None         # NOTE: VersionList != None and VersionList == None are always false
partition = VersionList()
for (nodejs_version, py_reqs) in data:
    # if a new partition is going to start, let's wrap up the just finished one
    if py_reqs != last_py_reqs:
        print("depends_on(\"python@{}\", when=\"@{}\", type=\"build\")".format(
            last_py_reqs,
            VersionRange(partition.lowest(), partition.highest())))
        partition = VersionList()

    last_py_reqs = py_reqs
    partition.add(nodejs_version)
```

Reply:
Python 3.8 and older are EOL and no longer supported, I don't think we even need to bother with lower bounds unless they are at least 3.10. So you might be fine only adding upper bounds.

Reply:
It's absolutely a good point. I'm not fully convinced it's the best we can do to remove

```python
# depends_on("python@2.7,3.5:3.7", when="@v13.0.0:v13.0.1", type="build")
# depends_on("python@2.7,3.5:3.8", when="@v13.1.0:v14.13.1", type="build")
```

becuase spack still knows some `python@3.6:3.8` version as deprecated. But I see that deprecated packages might not get full support.


So in the end, by "ignoring" EOL python versions we get

```python
depends_on("python@:3.9", when="@v14.14.0:v14.18.1", type="build")
depends_on("python@:3.10", when="@v14.18.2:v14.21.3", type="build")
depends_on("python@:3.9", when="@v15.0.0:v15.14.0", type="build")
depends_on("python@:3.9", when="@v16.0.0:v16.10.0", type="build")
depends_on("python@:3.10", when="@v16.11.0:v16.18.1", type="build")
depends_on("python@:3.11", when="@v16.19.0:v16.20.2", type="build")
depends_on("python@:3.10", when="@v17.0.0:v18.12.1", type="build")
depends_on("python@:3.11", when="@v18.13.0:v18.19.1", type="build")
depends_on("python@:3.12", when="@v18.20.0:v18.20.5", type="build")
depends_on("python@:3.10", when="@v19.0.0:v19.0.1", type="build")
depends_on("python@:3.11", when="@v19.1.0:v20.10.0", type="build")
depends_on("python@:3.12", when="@v20.11.0:v20.15.1", type="build")
depends_on("python@:3.13", when="@v20.16.0:v20.18.1", type="build")
depends_on("python@:3.11", when="@v21.0.0:v21.1.0", type="build")
depends_on("python@:3.12", when="@v21.2.0:v22.2.0", type="build")
depends_on("python@:3.13", when="@v22.3.0:v22.7.0", type="build")
```

I would go for this "extensive" list, and not for a filtered on known to spack version.

One last point. I'm going to mark next node-js versions as deprecated, since officially they can be built just with python versions not supported anymore (but still known to spack as deprecated).

```console
14.13.0    https://nodejs.org/dist/v14.13.0/node-v14.13.0.tar.gz
14.10.0    https://nodejs.org/dist/v14.10.0/node-v14.10.0.tar.gz
13.8.0     https://nodejs.org/dist/v13.8.0/node-v13.8.0.tar.gz
13.5.0     https://nodejs.org/dist/v13.5.0/node-v13.5.0.tar.gz
```

Reply:
I would still add upper bounds for these versions, even if they are deprecated. We can remove them in the next release.

Reply:
ah sorry, my bad. Then we are perfectly aligned, I fully agree with you. Let me fix that!